### PR TITLE
Configure Fastify Server with Custom Plugins + Fastify Upgrade

### PR DESCRIPTION
**Pull request by dthyresson**
This PR addresses the core issue noted in:

* https://github.com/redwoodjs/redwood/issues/4944
* https://github.com/redwoodjs/redwood/issues/5215

As @tbay06 noted:

> I've been using the server.config.js to play with some of the fastify settings, and I'm not sure how difficult/intensive it would be, but being able to enrich the fastify instance with some plugins via the config would be a really cool feature! Right now from what I can see the server.config.js exposes all of the [factory settings]

> (https://www.fastify.io/docs/latest/Reference/Server/#factory) for fastify, but if we were able to access the instance methods as well, that would open up a lot of awesome customization options.

Agreed!

> being able to enrich the fastify instance with some plugins via the config would be a really cool feature

The `server.config` now exposes a hook that configure plugins or routes per side.

See included docs for examples.

TODO:

* [x] Review with @dac09 
* [x] Add tests once functions finalized
* [x] Update docs for changes
* [x] figure out how to handle errors like duplicated plugin registration better
* [x] Check if we can use server.config.ts
* [x] Make sure we provide helpful errors OR backwards compatibility when using older server.config.js
* [x] codemod!

Note to @jtoar and @thedavidprice This is a breaking change and also should included as part of the Upgrade to Fastify v4 task here: https://github.com/redwoodjs/redwood/pull/5817

### CODE REVIEW

File: docs/docs/app-configuration-redwood-toml.md [hunk 1]
```diff
@@ -122,6 +125,138 @@ Using [redwood.toml's env var interpolation](#using-environment-variables-in-red
   serverConfig = "./api/${DEPLOY_ENVIRONMENT}-server.config.js"
 ```
 
+### Register Custom Fastify Plug-ins and Routes
+
+You can also register Fastify plugins and additional routes for the API and Web sides in the `configureFastifyForSide` function.
+
+This function has access to the Fastify instance and options, such as the sid (web, api, or proxy) that is being configured and other settings like the `apiRootPath` of the functions endpoint.
+
+ :::note
```
Comment (dac09) [inline docs/docs/app-configuration-redwood-toml.md:134]: Needs a space for it t owork I think

File: packages/api-server/src/__tests__/withFunctions.test.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    register: jest.fn(),
+    get: jest.fn(),
+    all: jest.fn(),
+    addContentTypeParser: jest.fn(),
+    log: console,
+  } as unknown as FastifyInstance
+
+  await withFunctions(mockedFastifyInstance, '.redwood/functions') // ?
+
+  const log = mockedFastifyInstance['log']['_buffer']
```
Comment (dac09) [inline packages/api-server/src/__tests__/withFunctions.test.ts:38]: I think these tests can probably made less flakey.

File: docs/docs/app-configuration-redwood-toml.md [hunk 3]
```diff
@@ -122,6 +134,120 @@ Using [redwood.toml's env var interpolation](#using-environment-variables-in-red
...
+
+And then configure them in your API side.
+
+Here, we setup compression to handle all requests (global), compress responses only if larger than 1K, and to prefer the `deflate` method over `gzip`.
+only allow one IP address to make 100 requests in a five minute window.
+
+Of course, these are just examples and please see the plugin documentaiton for the settings that best meet your needs.
+
+:::important
+You will need to install any custom plug-in packages to your project's `api` workspace.
```
Comment (dac09) [inline docs/docs/app-configuration-redwood-toml.md:179]:
```suggestion
You will need to install any custom plug-in packages to your project's `api` workspace.

To do this you run `yarn workspace api add @fastify/rate-limit @fastify/compress`
```

**Comment by dthyresson** [comment]
Note, no longer want to showcase route config, so updated title.

File: packages/api-server/src/requestHandlers/awsLambdaFastify.ts [hunk 4]
```diff
@@ -39,6 +48,57 @@ const lambdaEventForFastifyRequest = (
...
+    reply.headers(headersLowercase)
+  }
+}
+
+/**
+ * In case there are multi-value headers that are not in the headers object,
+ * we need to add them to the headers object and ensure the header names are lowercase
+ * and that if there is an array fo values, they are separated by a semi-colon.
+ */
+const mergeMultiValueHeaders = (
```
Comment (dac09) [inline packages/api-server/src/requestHandlers/awsLambdaFastify.ts:88]: Are you sure there are cases where multivalue headers are present and headers aren't? 

My understanding is that MVH are duplicated - its just a proxy setting.

File: packages/api-server/src/requestHandlers/awsLambdaFastify.ts [hunk 5]
```diff
@@ -39,6 +48,57 @@ const lambdaEventForFastifyRequest = (
...
+ *
+ * Just as in HTTP/1.x, header field names are strings of ASCII
+ * characters that are compared in a case-insensitive fashion.  However,
+ * header field names MUST be converted to lowercase prior to their
+ * encoding in HTTP/2.  A request or response containing uppercase
+ * header field names MUST be treated as malformed (Section 8.1.2.6).
+ *
+ * @see: https://tools.ietf.org/html/rfc2616#section-4.2
+ */
+const sanitizeHeaders = (
```
Comment (dac09) [inline packages/api-server/src/requestHandlers/awsLambdaFastify.ts:66]: As explained in the fastify issue - they've implemented this as a performance optimisation to only support lower case headers. Why do we need to sanitise them here? Just adds extra loops right?

**Review by dac09 — CHANGES_REQUESTED** [review]
Comments inline - not sure about MVH and sanitizeHeader changes

File: packages/api-server/src/requestHandlers/awsLambdaFastify.ts (hunk 4 above)
Comment (dac09) [reply packages/api-server/src/requestHandlers/awsLambdaFastify.ts:88]: Resolving as this is definitely required!

File: packages/api-server/src/requestHandlers/awsLambdaFastify.ts (hunk 5 above)
Comment (dac09) [reply packages/api-server/src/requestHandlers/awsLambdaFastify.ts:66]: DT will remove the PascalCase ContentType from our graphql function, and dbAuth.

**Comment by dthyresson** [comment]
> Comments inline - not sure about MVH and sanitizeHeader changes

No longer sanitizing headers as these will be standard lowercase in dbAuth and GraphQL Server.

Added tests for MVH and also in the Lambda request transform that MVH are converted into headers.

**Comment by dthyresson** [comment]
This PR also:

Updates api-server to latest Fastify v4.x

Main change was to the fastify-raw-body plugin that now requires an await.

See:

Note
You need to await the plugin registration to make sure the plugin is ready to use. All the routes defined before the plugin registration will be ignored. This change has been introduced in Fastify v4.
If not, then the body is empty and all GraphQL and other functions fail.

Includes

Fixes https://github.com/redwoodjs/redwood/pull/5718
Fixes https://github.com/redwoodjs/redwood/pull/5724
Fixes https://github.com/redwoodjs/redwood/pull/5723
Fixes https://github.com/redwoodjs/redwood/pull/5585

File: packages/codemods/src/codemods/v2.x.x/configureFastify/configureFastify.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import type { FileInfo, API, Identifier, MemberExpression } from 'jscodeshift'
+
+export default function transform(file: FileInfo, api: API) {
+  const j = api.jscodeshift
+  const ast = j(file.source)
+
+  ast.find(j.ExpressionStatement).forEach((path) => {
```
Comment (dac09) [inline packages/codemods/src/codemods/v2.x.x/configureFastify/configureFastify.ts:7]: Bit of a nitpick, so ok if you want to dismiss, but you could just do j.AssignmentExpression here... or even better if we used `j.ExportDeclaration

Comment (dac09) [reply packages/codemods/src/codemods/v2.x.x/configureFastify/configureFastify.ts:7]: Realised j.ExportDeclaration won't work here, but worth a check if we can make it more specific

**Comment by dthyresson** [comment]
@dac09 I think these final tests will soon pass and can merge.

**Comment by dac09** [comment]
Still seeing lint errors @dthyresson - unused imports and such. I'll fix it - but I think maybe your editor isn't showing these for you.

**Review by dac09 — APPROVED** [review]

**Pull request merged**